src/lib/branchpredictors/GenericPredictor.cc

```diff
@@ -68,30 +74,53 @@ BranchPrediction GenericPredictor::predict(uint64_t address, BranchType type,
     // Record that this address is a branch-and-link instruction
     rasHistory_[address] = 0;
   } else if (type == BranchType::Conditional) {
-    if (!prediction.taken) prediction.target = address + 4;
+    if (!prediction.isTaken) prediction.target = address + 4;
   }
+
+  // Store the hashed index for correct hashing in update()
+  ftq_.emplace_back(prediction.isTaken, hashedIndex);
+
+  // Speculatively update the global history
+  globalHistory_ =
+      ((globalHistory_ << 1) | prediction.isTaken) & globalHistoryMask_;
+
   return prediction;
 }
 
-void GenericPredictor::update(uint64_t address, bool taken,
-                              uint64_t targetAddress, BranchType type) {
-  // Get previous index calculated for the instruction address supplied
-  uint64_t hashedIndex = btbHistory_[address];
+void GenericPredictor::update(uint64_t address, bool isTaken,
+                              uint64_t targetAddress, BranchType type,
+                              uint64_t instructionId) {
+  // Make sure that this function is called in program order; and then update
+  // the lastUpdatedInstructionId variable
+  assert(instructionId >= lastUpdatedInstructionId_ &&
+         (lastUpdatedInstructionId_ = instructionId) >= 0 &&
+         "Update not called on branch instructions in program order");
```

Comment:
Unless a single branch instruction can call `update` multiple times, the initial condition should be `>` not `>=`

Reply:
A single branch instruction cannot call update multiple times.  It's >= because the first instruction will have instruction id = 0, which is what it starts lastUpdatedInstructionId_ -- I wanted to keep the same data type as the instructionId, which is unsigned so I couldn't start below 0.  Happy to change to a better solution.